src/rgw/rgw_auth_s3.h

```diff
@@ -32,6 +32,54 @@ static constexpr auto RGW_AUTH_GRACE = std::chrono::minutes{15};
 // returns true if the request time is within RGW_AUTH_GRACE of the current time
 bool is_time_skew_ok(time_t t);
 
+class STSAuthStrategy : public rgw::auth::Strategy,
+                        public rgw::auth::RemoteApplier::Factory,
+                        public rgw::auth::LocalApplier::Factory {
```

Comment:
I have a separate auth strategy for STS because, it needs to be able to authenticate users which have been authenticated by our LocalEngine (in case of AssumeRole) and users which have been authenticated by our RemoteEngines (Keystone/ LDAP in case of GetSessionToken) and hence have the different appliers pass down the permissions which are allowed by AssumeRole (Role policies) and GetSessionToken(Perm Mask etc). Hence STSEngine needs to access both LocalApplier and RemoteApplier. Both ExternalAuthStrategy and LocalAuthStrategy access RemoteApplier and LocalApplier resepectively and not both, hence I couldn't find an appropriate place to insert STSEngine in these auth strategies.

Reply:
@cbodley @adamemerson @mattbenjamin @mdw-at-linuxbox : Please review

Reply:
Doc added